ddht/v5/crawl.py

```diff
@@ -0,0 +1,234 @@
+
+                await trio.sleep(1)  # don't overburden this peer
+        except StopScanning:
+            pass
+        finally:
+            # we only send one packet per peer. do some cleanup now or we'll leak memory.
+            self.client.discard_peer(remote_enr.node_id)
+
+    async def read_from_queue_until_done(self) -> None:
+        while True:
+
+            # CAUTION: Do not insert any code here. There must not be any checkpoints
+            #          between leaving active_tasks (in the previous loop iteration) and
+            #          trying to read from the queue.
+
+            try:
+                enr = self.enrqueue_recv.receive_nowait()
+            except trio.WouldBlock:
+                if len(self.active_tasks) > 0:
+                    # some tasks are still active so it's too soon to quit. Instead, block
+                    # until new work comes in. If we've truly run out of work then this
+                    # will block forever but that's okay, another task will notice and
+                    # cause our cancellation.
+                    enr = await self.enrqueue_recv.receive()
+                else:
+                    # nobody else is performing any work and there's also no work left to
+                    # be done. It's time to quit! Calling this causes all other tasks to
+                    # be canceled.
+                    await self.manager.stop()
+                    return
+
+            # CAUTION: Do not insert any code here. There must not be any checkpoints
+            #          between popping from the queue and entering active_tasks.
+
+            with self.active_tasks.enter():
+                await self.scan_enr_buckets(enr)
+
+    async def schedule_enr_to_be_visited(self, enr: ENRAPI) -> None:
+        if enr.node_id in self.seen_nodeids:
+            # In order to be nicer on the network only send a single packet to each
+            # remote node.
+            return
+
+        if IP_V4_ADDRESS_ENR_KEY not in enr:
+            logger.info(f"dropping ENR without IP address enr={enr} kv={enr.items()}")
+            return
+
+        self.seen_nodeids.add(enr.node_id)
+
+        try:
+            self.enr_db.set_enr(enr)
+        except OldSequenceNumber:
+            logger.info(f"dropping old ENR. enr={enr} kv={enr.items()}")
+            return
+
+        enqueued = self.enrqueue_send.statistics().current_buffer_used
+        logger.info(
+            f"found ENR. count={len(self.seen_nodeids)} enr={enr} node_id={enr.node_id.hex()}"
+            f" sequence_number={enr.sequence_number} kv={enr.items()} queue_len={enqueued}"
+        )
+
+        await self.enrqueue_send.send(enr)
+
+    async def run(self) -> None:
+        logger.info("crawling!")
+
+        boot_info = self._boot_info
+
+        if boot_info.is_upnp_enabled:
+            logger.info(
+                "uPNP will not be used; crawling does not require listening for "
+                "incoming connections."
```

Comment:
Does UDP somehow magically punch through?  Even though we don't listen for connections we do have to receive responses to our messages.  Mostly just curious how this works because it confuses me.

Reply:
It sure does! I'm going off memory so might be getting this wrong, but this works in the same way that you can open TCP connections to sites like google.com without needing to punch any holes for them. When you send an outbound packet the NAT remembers the port number you used and for a short period of time forwards packets to that port to you.

If I'm reading [this](https://tools.ietf.org/html/rfc5382) right the port stays open for at least 4 minutes.